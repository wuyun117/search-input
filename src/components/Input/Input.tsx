import React from "react";
import PropTypes from "prop-types";
import "./Input.scss";

/**
 * <Input
 *   className="MyInput"
 *   data-something="Value"
 *   value="Hello, World!"
 *   onChange={(value) => console.log('You typed', value)}
 * />
 *
 * @prop {string} value The default value for the input.
 * @prop {string} placeholder The placeholder text.
 * @prop {Function} onChange Callback that will receive current input value.
 * @prop {mixed} ... All other props will be forwarded to the native DOM input.
 */

interface IInputProps {
  id?: string;
  className?: string;
  placeholder?: string;
  value: string;
  onChange: (arg: string) => void;
  onKeyUp: (arg1: string, arg2: number, ) => void;
}

const Input: React.FC<IInputProps> = (props: IInputProps) => {

  const { className, value, onChange, onKeyUp, ...otherProps } = props;

  return (
    <input
      className={"Input " + (className || "")}
      type="text"
      value={value}
      data-testid="search-input"
      onChange={event => onChange(event.target.value)}
      onKeyUp={event => onKeyUp(event.key, event.keyCode)}
      {...otherProps}
    />
  );
}

Input.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  placeholder: PropTypes.string,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onKeyUp: PropTypes.func.isRequired
}

export default Input;
