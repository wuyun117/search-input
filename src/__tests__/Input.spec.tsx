import React from "react";
import { render, act, fireEvent } from "@testing-library/react";

import Input from "../components/Input/Input";

describe("Button component", () => {
  it("component render correctly", () => {
    const changeFn = jest.fn();
    const keyUpFn = jest.fn();
    const wrapper = render(<Input value="Syd" placeholder="Search Input" onChange={changeFn} onKeyUp={keyUpFn} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should have placeholder: Search Input and value: Sydney", async () => {
    const changeFn = jest.fn();
    const keyUpFn = jest.fn();
    const { getByTestId } = render(<Input value="Sydney" placeholder="Search Input" onChange={changeFn} onKeyUp={keyUpFn} />);
    expect(getByTestId("search-input")).toHaveAttribute('placeholder', 'Search Input');
    expect(getByTestId("search-input")).toHaveValue('Sydney');
  });

  it("should trigger onclick event", async () => {
    const changeFn = jest.fn();
    const keyUpFn = jest.fn();
    const value = "Chatswood";
    const { getByTestId } = render(<Input value="Sydney" placeholder="Search Input" onChange={changeFn} onKeyUp={keyUpFn} />);
    expect(getByTestId("search-input")).toHaveValue('Sydney');
    await act(async () => {
      fireEvent.change(getByTestId("search-input"), {
        target: { value },
      });
    });
    expect(changeFn).toHaveBeenLastCalledWith(value);
  });

  it("should trigger onKeyUp event", async () => {
    const changeFn = jest.fn();
    const keyUpFn = jest.fn();
    const event = {
      key: 'ArrowUp',
      keyCode: 38
    }
    const { getByTestId } = render(<Input value="Sydney" placeholder="Search Input" onChange={changeFn} onKeyUp={keyUpFn} />);
    expect(getByTestId("search-input")).toHaveValue('Sydney');
    await act(async () => {
      fireEvent.keyUp(getByTestId("search-input"), {
        ...event
      });
    });
    expect(keyUpFn).toHaveBeenLastCalledWith(event.key, event.keyCode);
  });
});