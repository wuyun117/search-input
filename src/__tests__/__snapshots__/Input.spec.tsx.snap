// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component component render correctly 1`] = `
<div>
  <input
    class="Input "
    data-testid="search-input"
    placeholder="Search Input"
    type="text"
    value="Syd"
  />
</div>
`;
